src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEReader.EmbeddedPortablePdb.cs

```diff
@@ -83,44 +81,63 @@ internal static unsafe ImmutableArray<byte> DecodeEmbeddedPortablePdbDebugDirect
 
             try
             {
-                decompressed = new byte[decompressedSize];
+                decompressed = new NativeHeapMemoryBlock(decompressedSize);
             }
             catch
             {
                 throw new BadImageFormatException(SR.DataTooBig);
             }
 
-            var compressed = new ReadOnlyUnmanagedMemoryStream(headerReader.CurrentPointer, headerReader.RemainingBytes);
-            var deflate = new DeflateStream(compressed, CompressionMode.Decompress, leaveOpen: true);
-
-            if (decompressedSize > 0)
+            bool success = false;
+            try
             {
-                int actualLength;
+                var compressed = new ReadOnlyUnmanagedMemoryStream(headerReader.CurrentPointer, headerReader.RemainingBytes);
+                var deflate = new DeflateStream(compressed, CompressionMode.Decompress, leaveOpen: true);
 
-                try
-                {
-                    actualLength = deflate.TryReadAll(decompressed, 0, decompressed.Length);
-                }
-                catch (InvalidDataException e)
+                if (decompressedSize > 0)
                 {
-                    throw new BadImageFormatException(e.Message, e.InnerException);
+                    int actualLength;
+
+                    try
+                    {
+#if NETCOREAPP3_0_OR_GREATER
+                        actualLength = deflate.Read(new Span<byte>(decompressed.Pointer, decompressed.Size));
```

Comment:
This isn't guaranteed to read Size bytes; it could end up reading fewer.

Reply:
Fixed. This is unfortunate though. Seems like it would be worth it to add `ReadAll` method on Stream. We have one in Roslyn, we have one in SRM, etc. Seems like everyone is gonna have one.

Reply:
https://github.com/dotnet/runtime/issues/16598